File: rewrite/web/index/collection.js
import { Pieces } from '../pieceDefs.js'

export function collectionImage(index) { 
    return '/web/img/pick_'+Pieces[window.Collection[index]]+'.png'
}

export function pickImage(kind) { return '/web/img/pick_'+Pieces[kind]+'.png' }

export const LeftPick = -1
export const RightPick = -2
export const NotInCollection = 0

export function collectionSelector(slot) {
    switch (slot) {
    case NotInCollection:
        return ''
    case LeftPick:
        return '#leftpick'
    case RightPick:
        return '#rightpick'
    }
    return '#c' + (slot-1)
}

File: rewrite/collection.go
package main

import "github.com/pciet/wichess/rules"

const (
	CollectionCount = 21

	NotInCollection CollectionSlot = 0

	// The two random picks are considered part of the player's collection.
	LeftPick  CollectionSlot = -1
	RightPick CollectionSlot = -2
)

type (
	Collection [CollectionCount]Piece

	// The CollectionSlot is the array index into Collection plus one to be a human readable value.
	CollectionSlot int8

	AddressedCollectionSlot struct {
		Slot          CollectionSlot `json:"s"`
		rules.Address `json:"a"`
	}

	RandomPicks struct {
		Left  rules.PieceKind
		Right rules.PieceKind
	}
)

func (a Collection) Kinds() [CollectionCount]rules.PieceKind {
	var out [CollectionCount]rules.PieceKind
	for i, p := range a {
		out[i] = p.Kind
	}
	return out
}

func (a CollectionSlot) Int() int { return int(a) }